hazelcast/src/main/java/com/hazelcast/map/impl/query/MapLocalQueryRunner.java

```diff
@@ -0,0 +1,317 @@
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.exception.RetryableHazelcastException;
+import com.hazelcast.spi.partition.IPartitionService;
+import com.hazelcast.util.Clock;
+import com.hazelcast.util.IterationType;
+import com.hazelcast.util.executor.ManagedExecutorService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+import static com.hazelcast.map.impl.query.MapQueryEngineUtils.newQueryResult;
+import static com.hazelcast.map.impl.query.MapQueryEngineUtils.newQueryResultForSinglePartition;
+import static com.hazelcast.query.PagingPredicateAccessor.getNearestAnchorEntry;
+import static com.hazelcast.spi.ExecutionService.QUERY_EXECUTOR;
+import static com.hazelcast.util.SortingUtil.compareAnchor;
+import static com.hazelcast.util.SortingUtil.getSortedSubList;
+
+/**
+ * Runs query operations in the calling thread (thus blucking it)
+ * Query evaluation per partition is run in a sequential fashion.
+ * <p>
+ * Used by query operations only: QueryOperation & QueryPartitionOperation
+ * Should not be used by proxies or any other query related objects.
+ */
+public class MapLocalQueryRunner {
+
+    protected final MapServiceContext mapServiceContext;
+    protected final NodeEngine nodeEngine;
+    protected final ILogger logger;
+    protected final QueryResultSizeLimiter queryResultSizeLimiter;
+    protected final InternalSerializationService serializationService;
+    protected final IPartitionService partitionService;
+    protected final QueryOptimizer queryOptimizer;
+    protected final OperationService operationService;
+    protected final ClusterService clusterService;
+    protected final LocalMapStatsProvider localMapStatsProvider;
+    protected final ManagedExecutorService executor;
+
+    public MapLocalQueryRunner(MapServiceContext mapServiceContext, QueryOptimizer optimizer) {
+        this.mapServiceContext = mapServiceContext;
+        this.nodeEngine = mapServiceContext.getNodeEngine();
+        this.serializationService = (InternalSerializationService) nodeEngine.getSerializationService();
+        this.partitionService = nodeEngine.getPartitionService();
+        this.logger = nodeEngine.getLogger(getClass());
+        this.queryResultSizeLimiter = new QueryResultSizeLimiter(mapServiceContext, logger);
+        this.queryOptimizer = optimizer;
+        this.operationService = nodeEngine.getOperationService();
+        this.clusterService = nodeEngine.getClusterService();
+        this.localMapStatsProvider = mapServiceContext.getLocalMapStatsProvider();
+        this.executor = nodeEngine.getExecutionService().getExecutor(QUERY_EXECUTOR);
+    }
+
+    // full query = index query (if possible), then partition-scan query
+    public QueryResult runFullQuery(String mapName, Predicate predicate, IterationType iterationType)
```

Comment:
Probably so. I'll be adding fast-aggregations and projections in the next couple of days, thus I would not like to optimize the API prematurely. Adding those will be a great validation of the API and will definitely result in its changes. Other hints appreciated.